src/client/datascience/editor-integration/codewatcher.ts

```diff
@@ -982,6 +972,56 @@ export class CodeWatcher implements ICodeWatcher {
         }
     }
 
+    private createAddCodePromise(
+        code: string,
+        file: Uri,
+        line: number,
+        leftCount: number,
+        editor?: TextEditor,
+        debug?: boolean
+    ): Promise<boolean> {
+        // Get our add code promise and tack on a then for getting the next queue item
+        const addCodePromise = this.addCode(code, file, line, editor, debug);
+        const addCodePromiseEdit = addCodePromise.then(async (result) => {
+            this.executingAddCode = undefined;
+
+            if (result) {
+                // Success. Check queue for next
+                const nextCreationFunction = this.addCodeQueue.shift();
+
+                if (nextCreationFunction) {
+                    this.executingAddCode = nextCreationFunction();
+                }
+            } else {
+                // Failure. Print our failure message and clear the queue
+                await this.addErrorMessage(file, leftCount);
+                this.addCodeQueue = [];
+            }
+
+            return result;
+        });
+
+        return addCodePromiseEdit;
+    }
+
+    // Queue up our addCode promises so that we run them in order
+    private queueAddCode(
+        code: string,
+        file: Uri,
+        line: number,
+        leftCount: number,
+        editor?: TextEditor,
+        debug?: boolean
+    ) {
+        // Either start up this addCode, or queue it for later
+        if (this.executingAddCode) {
```

Comment:
I feel like TS / JS has a bunch of way to do this type of queuing and they all always look a bit janky to me. Speak up if it looks like I'm doing something unreasonable here, but this way felt the cleanest to me. (Had a promise chain thing I was initially working with that felt hard to read / debug).

Reply:
I think this is a bit of an overkill. It could just chain the promise. Like the interactive window does itself. 

Essentially you have a single promise - addCodePromise that you init to a resolved promise.

addCode changes to do something like so:

```
        // Chain execution promises so that cells are executed in the right order
        if (this.addCodePromise) {
            this.addCodePromise= this.addCodePromise.then(() =>
                this.addCodeImpl(code, fileUri, line, isDebug)
            );
        } else {
            this.addCodePromise= this.addCodeImpl(code, fileUri, line, isDebug);
        }
        try {
            return await this.addCodePromise;
        } catch (exc) {
            // Rethrow, but clear promise so we can execute again
            this.addCodePromise= undefined;
            throw exc;
        }
```

Old addCode method becomes addCodeImpl

Actually feels like we could have a method/service to do this cause we do this a lot.

Reply:
This would also leave the old code the same as the behavior wouldn't have to account for a queue.

Reply:
Trying to think of what that generic service might be. 

It would look like a function. A function which returns a promise. 

I wonder if we could do it with a decorator. That would be even better. Something like so:

```
@chainable
public addCode(...): Promise 
```

Where the chainable decorator would do the code I pasted above.

Reply:
I agree, and would suggest moving all of that into the existing `addCode` method, unless i'm reading it incorrectly

Reply:
I suggest the addCode method as that the choke point, thats what everything calls today.
Perhaps rename from `addCode` to `queueAddCode` or the like and make it sync

Reply:
Then we could use that in both the interractive window and here (and probably ten other places where we chain stuff)

Reply:
I had initially started with the promise chain starting from a resolved promise. Felt like it was getting icky with the promises resolving to false (instead of throwing) and the error handler function having to get tacked on. But I'm not adverse to going back. Let me restart from there and see if I can get it looking better.

Reply:
I tried this on the current interactive window 'submitCodeImpl' and it seems to work. Then we can use this to do promise chaining anywhere:

```typescript
// Turn a method that returns a promise into something that uses a chain of promises
export function chainPromise() {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function (_target: Object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>) {
        const originalMethod = descriptor.value;
        const chainedKey = `chained${propertyKey}`;

        // eslint-disable-next-line , @typescript-eslint/no-explicit-any
        descriptor.value = function (...args: any[]) {
            // Check for promise in the object.
            let currentValue = (this as any)[chainedKey] as Promise<any>;
            if (currentValue) {
                currentValue = currentValue.then(() => originalMethod.apply(this, args));
            } else {
                currentValue = originalMethod.apply(this, args);
            }

            // Save promise in object
            (this as any)[chainedKey] = currentValue;

            // If promise fails, clear it.
            return currentValue
                .then((r) => r)
                .catch((e) => {
                    (this as any)[chainedKey] = undefined;
                    throw e;
                });
        };

        return descriptor;
    };
}
```

This changes submitCodeImpl (or your addCode) to this instead:

```
@chainPromise()
    private async submitCodeImpl(code: string, fileUri: Uri, line: number, isDebug: boolean) {
    // Internals don't worry about chaining anymore
}
```

Reply:
The decorator I didn't like in this situation. Just in general I find them hard to debug (prefer them much more for extra stuff like telemetry / timing on a function as opposed to something like this) and the differences from the usage somewhere like submitCodeImpl made it hard to work in both situations.

Reply:
Hmm, thought I put a comment here. I didn't like the decorator for this (though your implementation is rad clever) it feels like really core functionality to be putting into a decorator (versus timing / telemetry) and not easy to debug and the mild differences between something like the addCode here and submitCode in the interactive window (things like the false return value handling) didn't lend to sharing one decorator.